models/user/user.go

```diff
@@ -316,37 +316,94 @@ func (u *User) GenerateEmailActivateCode(email string) string {
+func GetUserFollowers(ctx context.Context, opts GetUserFollowOptions) ([]*User, int64, error) {
+	sess := db.GetEngine(ctx).Where(opts.toCond(true)).
 		Join("LEFT", "follow", "`user`.id=follow.user_id")
 
-	if listOptions.Page != 0 {
-		sess = db.SetSessionPagination(sess, &listOptions)
+	if opts.Page != 0 {
+		sess = db.SetSessionPagination(sess, &opts.ListOptions)
 
-		users := make([]*User, 0, listOptions.PageSize)
-		return users, sess.Find(&users)
+		users := make([]*User, 0, opts.PageSize)
+		count, err := sess.FindAndCount(&users)
+		return users, count, err
 	}
 
 	users := make([]*User, 0, 8)
-	return users, sess.Find(&users)
+	count, err := sess.FindAndCount(&users)
+	return users, count, err
+}
+
+// GetFeedsOptions options for getting the user's followers or followings.
+type GetUserFollowOptions struct {
+	db.ListOptions
+	Actor         *User // the user viewing the followings
+	RequestedUser *User // the user we want followings for
+}
+
+func (opts GetUserFollowOptions) toCond(checkFollowers bool) builder.Cond {
+	cond := builder.NewCond()
+
+	if checkFollowers {
+		cond = cond.And(builder.Eq{"`follow`.follow_id": opts.RequestedUser.ID})
+	} else {
+		cond = cond.And(builder.Eq{"`follow`.user_id": opts.RequestedUser.ID})
+	}
+
+	// If the actor is not signed in. Only show users that have their visibility
+	// set to public.
+	if opts.Actor == nil {
+		return cond.And(builder.Eq{"`user`.visibility": structs.VisibleTypePublic})
+	}
+
+	// Fast path for admins.
+	if opts.Actor.IsAdmin {
+		return cond
+	}
+
+	// If the actor is signed in, then we allow all limited & public accounts to be seen.
+	// However the actor can also see private accounts if they have are in the same organisation
+	// as that user.
+	cond = cond.And(
+		builder.Or(
+			// Include all limited & public accounts.
+			builder.Lte{"`user`.visibility": structs.VisibleTypeLimited},
+			// Check which private users can be included.
+			// Get the actor's orginisations and check if the private
+			// users are in those orgs.
+			builder.And(
+				// Specify all private users.
+				builder.Eq{"`user`.visibility": structs.VisibleTypePrivate},
+				// Is the user's id in the organisation that the actor is in?
+				builder.In("`user`.id",
+					builder.Select("uid").From("org_user").Where(
+						builder.In("org_id",
+							// Get the actor's orginaisations.
+							builder.Select("org_id").From("org_user").Where(builder.Eq{"uid": opts.Actor.ID})),
+					)),
+			),
+		),
+	)
+
+	return cond
 }
 
 // GetUserFollowing returns range of user's following.
-func GetUserFollowing(u *User, listOptions db.ListOptions) ([]*User, error) {
-	sess := db.GetEngine(db.DefaultContext).
-		Where("follow.user_id=?", u.ID).
-		Join("LEFT", "follow", "`user`.id=follow.follow_id")
+func GetUserFollowing(ctx context.Context, opts GetUserFollowOptions) ([]*User, int64, error) {
+	sess := db.GetEngine(ctx).
+		Join("LEFT", "follow", "`user`.id=follow.follow_id").
```

Comment:
Shouldn't these be INNER JOINs instead of LEFT JOINs? In this case it does not change the result because of the WHERE clause.

Reply:
I'm just copying the previous code. I really don't want to make any more changes that doesn't cover the scope of the PR. It just leads to regressions and problems and arguments that I don't feel like fighting. But yeah it seems you're right we should only target followers that correspond to a existent user in the database, do you want to make a separate pull request for it?

Reply:
Yes, I will create a PR.